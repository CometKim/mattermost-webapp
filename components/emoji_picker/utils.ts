// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {
    Emoji as _Emoji,
    SystemEmoji,
    CustomEmoji,
} from 'mattermost-redux/types/emojis';

export type EmojiCategory = (
    | 'people'
    | 'nature'
    | 'foods'
    | 'activity'
    | 'places'
    | 'objects'
    | 'symbols'
    | 'flags'
    | 'custom'
);

export type Emoji = _Emoji;

export function isSystemEmoji(emoji: Emoji): emoji is SystemEmoji {
    return 'batch' in emoji;
}

export function isCustomEmoji(emoji: Emoji): emoji is CustomEmoji {
    return 'name' in emoji;
}

export function getEmojiName(emoji: Emoji) {
    return isSystemEmoji(emoji) ? emoji.aliases[0] : emoji.name;
}
