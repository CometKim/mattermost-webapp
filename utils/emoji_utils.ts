
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Emoji} from 'mattermost-redux/types/emojis';
import {isCustomEmoji} from 'mattermost-redux/utils/emoji_utils';

const defaultRule = (aName: string, bName: string, emojiA: Emoji, emojiB: Emoji) => {
    if (emojiA && emojiB && emojiA.category === 'custom' && emojiB.category !== 'custom') {
        return 1;
    } else if (emojiB && emojiA && emojiB.category === 'custom' && emojiA.category !== 'custom') {
        return -1;
    }

    return aName.localeCompare(bName);
};

const thumbsDownRule = (otherName: string) => {
    if (otherName === 'thumbsup' || otherName === '+1') {
        return 1;
    }
    return 0;
};

const thumbsUpRule = (otherName: string) => {
    if (otherName === 'thumbsdown' || otherName === '-1') {
        return -1;
    }
    return 0;
};

const customRules = {
    thumbsdown: thumbsDownRule,
    '-1': thumbsDownRule,
    thumbsup: thumbsUpRule,
    '+1': thumbsUpRule,
} as { [key: string]: (name: string) => number };

export function getEmojiName(emoji: Emoji, searchedName?: string) {
    // There's an edge case for custom emojis that start with a thumb prefix.
    // It doesn't match the first alias for the relevant system emoji.
    // We don't have control over the names or aliases of custom emojis...
    // ... and how they compare to the relevant system ones.
    // So we need to search for a matching alias in the whole array.
    // E.g. thumbsup-custom vs [+1, thumbsup]

    if (isCustomEmoji(emoji)) {
        return emoji.name;
    }

    if (searchedName) {
        return emoji.aliases.find((alias) => alias.startsWith(searchedName)) || emoji.aliases[0];
    }

    return emoji.aliases[0];
}

export function compareEmojis(emojiA: Emoji, emojiB: Emoji, searchedName?: string) {
    const aName = getEmojiName(emojiA, searchedName);
    const bName = getEmojiName(emojiB, searchedName);

    // Have the emojis that contain the search appear first
    const aPrefix = aName.startsWith(searchedName || '');
    const bPrefix = bName.startsWith(searchedName || '');

    if (aPrefix === bPrefix) {
        if (customRules[aName]) {
            return customRules[aName](bName) || defaultRule(aName, bName, emojiA, emojiB);
        }

        return defaultRule(aName, bName, emojiA, emojiB);
    } else if (aPrefix) {
        return -1;
    }

    return 1;
}
