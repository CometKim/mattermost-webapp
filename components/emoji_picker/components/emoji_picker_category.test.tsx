// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {EmojiCategory} from '../utils';

import EmojiPickerCategory, {EmojiPickerCategoryProps} from './emoji_picker_category';

describe('components/EmojiPicker/components/EmojiPickerCategory', () => {
    const baseProps: EmojiPickerCategoryProps = {
        category: 'people',
        // eslint-disable-next-line react/jsx-no-literals
        selected: true,
        enable: true,
    };

    it.each([
        'activity',
        'custom',
        'flags',
        'foods',
        'nature',
        'objects',
        'people',
        'places',
        'recent',
        'symbols',
    ])('should match snapshot for %s', (category) => {
        const wrapper = shallow(
            <EmojiPickerCategory
                {...baseProps}
                category={category as EmojiCategory}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot when selected=false', () => {
        const wrapper = shallow(
            <EmojiPickerCategory
                {...baseProps}
                selected={false}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot when enable=false', () => {
        const wrapper = shallow(
            <EmojiPickerCategory
                {...baseProps}
                enable={false}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
