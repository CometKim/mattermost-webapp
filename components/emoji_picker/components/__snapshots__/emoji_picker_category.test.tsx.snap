// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for activity 1`] = `
<button
  aria-label="activity"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Activity />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for custom 1`] = `
<button
  aria-label="custom"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Custom />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for flags 1`] = `
<button
  aria-label="flags"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Flags />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for foods 1`] = `
<button
  aria-label="foods"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Foods />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for nature 1`] = `
<button
  aria-label="nature"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Nature />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for objects 1`] = `
<button
  aria-label="objects"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Objects />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for people 1`] = `
<button
  aria-label="people"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <People />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for places 1`] = `
<button
  aria-label="places"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Places />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for recent 1`] = `
<button
  aria-label="recent"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Recent />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot for symbols 1`] = `
<button
  aria-label="symbols"
  className="emoji-picker__category emoji-picker__category--selected"
>
  <Symbols />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot when enable=false 1`] = `
<button
  aria-label="people"
  className="emoji-picker__category emoji-picker__category--selected disable"
>
  <People />
</button>
`;

exports[`components/EmojiPicker/components/EmojiPickerCategory should match snapshot when selected=false 1`] = `
<button
  aria-label="people"
  className="emoji-picker__category"
>
  <People />
</button>
`;
